### PR TITLE
Import tool optimizations and less littering

### PR DESCRIPTION
**Pull request by tinwelint**
- Removes need for total ordering of batches, which means no waiting/coordination => 10-15% performance improvement.
- Removes lots of List/Iterator instantiations happening in and around PropertyRecord

### CODE REVIEW

File: community/kernel/src/main/java/org/neo4j/kernel/impl/store/record/PropertyBlock.java [hunk 1]
```diff
@@ -70,17 +71,29 @@ public void setSingleBlock( long value )
...
+        if ( valueRecords == null )
+        {
+            valueRecords = new LinkedList<>();
+        }
         valueRecords.add( record );
     }
 
+    public void setValueRecords( List<DynamicRecord> valueRecords )
+    {
+        this.valueRecords = valueRecords;
```
Comment (davidegrohmann) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/store/record/PropertyBlock.java]: perhaps you want to assert that `this.valueRecords` is null here?

File: community/kernel/src/main/java/org/neo4j/kernel/impl/store/record/PropertyRecord.java [hunk 2]
```diff
@@ -99,37 +103,78 @@ public long getRelId()
...
+        {
+            throw new NoSuchElementException();
+        }
+        return blockRecords[blockRecordsIteratorCursor++];
+    }
+
+    @Override
+    public void remove()
+    {
+        if ( blockRecordsIteratorCursor == 0 )
```
Comment (davidegrohmann) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/store/record/PropertyRecord.java]: shouldn't this be `blockRecordsCursor == 0` instead?

File: community/kernel/src/main/java/org/neo4j/unsafe/impl/batchimport/cache/idmapping/string/EncodingIdMapper.java [hunk 3]
```diff
@@ -82,7 +82,9 @@
     private boolean readyForUse;
     private long[][] sortBuckets;
     private long size;
-    private final List<IdGroup> idGroups = new ArrayList<>();
+
+    private final IdGroup[] idGroups = new IdGroup[10];
```
Comment (davidegrohmann) [inline community/kernel/src/main/java/org/neo4j/unsafe/impl/batchimport/cache/idmapping/string/EncodingIdMapper.java]: Where is `10` coming from?

File: community/kernel/src/test/java/org/neo4j/unsafe/impl/batchimport/input/csv/CsvInputTest.java [hunk 4]
```diff
@@ -640,7 +640,10 @@ private void assertRelationship( InputRelationship relationship, long id,
             Group endNodeGroup, Object endNode,
             String type, Object[] properties )
     {
-        assertEquals( id, relationship.id() );
+        if ( relationship.hasSpecificId() )
+        {
+            assertEquals( id, relationship.specificId() );
+        }
```
Comment (davidegrohmann) [inline community/kernel/src/test/java/org/neo4j/unsafe/impl/batchimport/input/csv/CsvInputTest.java]: perhaps we need an else case here checking the that given `id` is `-1`?

**Comment by davidegrohmann** [comment]
It would be nice to have some extra test to check the `iterator` implementation in `PropertyRecord`.

**Comment by tinwelint** [comment]
Thanks @davidegrohmann I'll address your comments above

File: community/kernel/src/main/java/org/neo4j/kernel/impl/store/record/PropertyRecord.java (hunk 2 above)
Comment (tinwelint) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/store/record/PropertyRecord.java]: This `remove()` method comes from `Iterator` and so the item to be removed is the one we just iterator to in `next()`. `blockRecordsIteratorCursor` is incremented in `next()` and so we know that we can remove the "current iterator item" if we're not right at the beginning.

Although the current implementation breaks the `Iterator#contract` in that two consecutive calls to `remove()` might succeed and remove two items. There has to be a call to `next` in between. I'll fix that.

File: community/kernel/src/main/java/org/neo4j/unsafe/impl/batchimport/cache/idmapping/string/EncodingIdMapper.java (hunk 3 above)
Comment (tinwelint) [inline community/kernel/src/main/java/org/neo4j/unsafe/impl/batchimport/cache/idmapping/string/EncodingIdMapper.java]: Just an arbitrary number that most imports will be within range of. I meant to have it grow on demand in `put`, but forgot. I'll fix that

File: community/kernel/src/test/java/org/neo4j/unsafe/impl/batchimport/input/csv/CsvInputTest.java (hunk 4 above)
Comment (tinwelint) [inline community/kernel/src/test/java/org/neo4j/unsafe/impl/batchimport/input/csv/CsvInputTest.java]: Actually `CsvInput` will always produce `InputRelationship`s w/o specific ids. I'll do an `assertFalse( relationship.hasSpecificId() )` instead there.

**Comment by tinwelint** [comment]
@davidegrohmann are you happy with those changes?

**Pull request merged**